src/provider/mod.rs

```diff
@@ -75,13 +75,40 @@ pub const DEFAULT_BIND_ADDR: ([u8; 4], u16) = ([127, 0, 0, 1], 4433);
 /// The returned [`Provider`] is awaitable to know when it finishes.  It can be terminated
 /// using [`Provider::shutdown`].
 #[derive(Debug)]
-pub struct Builder<E: ServiceEndpoint<ProviderService> = DummyServerEndpoint> {
+pub struct Builder<E = DummyServerEndpoint, C = ()>
+where
+    E: ServiceEndpoint<ProviderService>,
+    C: CustomGetHandler,
+{
     bind_addr: SocketAddr,
     keypair: Keypair,
     auth_token: AuthToken,
     rpc_endpoint: E,
     db: Database,
     keylog: bool,
+    custom_get_handler: C,
+}
+
+/// A custom get request handler that allows the user to make up a get request
+/// on the fly.
+pub trait CustomGetHandler: Send + Sync + Clone + 'static {
+    /// Handle the custom request, given an opaque data blob from the requester.
+    fn handle(
+        &self,
+        request: Bytes,
+        db: Database,
+    ) -> BoxFuture<'static, anyhow::Result<GetRequest>>;
```

Comment:
did you consider a version where this is responsible for directly feeding you the collection and blob data?

With this design you have to add the collection to the db, construct the request and then magically know when the response has been sent and remove the stuff from the db again if that's what you need.

Reply:
Yes, I think we need something where the temp collection gets removed automatically after the request gets processed.

But I think @dignifiedquire said that he would prefer this to be merged more or less as-is so appa can depend on it, and we make it nicer later after we did hole punching stuff.